src/main/kotlin/io/sc3/peripherals/posters/printer/PosterPrinterPeripheral.kt

```diff
@@ -0,0 +1,111 @@
+package io.sc3.peripherals.posters.printer
+
+import dan200.computercraft.api.lua.LuaException
+import dan200.computercraft.api.lua.LuaFunction
+import dan200.computercraft.api.lua.LuaValues.*
+import dan200.computercraft.api.lua.MethodResult
+import dan200.computercraft.api.lua.MethodResult.of
+import dan200.computercraft.api.peripheral.IComputerAccess
+import dan200.computercraft.api.peripheral.IPeripheral
+import io.sc3.peripherals.posters.PosterPrintData
+import io.sc3.peripherals.prints.MAX_LABEL_LENGTH
+import io.sc3.peripherals.prints.MAX_TOOLTIP_LENGTH
+import io.sc3.peripherals.prints.printer.PrinterBlockEntity
+import java.util.*
+
+class PosterPrinterPeripheral(val be: PosterPrinterBlockEntity) : IPeripheral {
+  override fun getType() = "poster_printer"
+
+  @LuaFunction(mainThread = true)
+  fun reset() {
+    be.data = PosterPrintData()
+    be.dataUpdated()
+  }
+
+  @LuaFunction(mainThread = true)
+  fun getLabel(): String? = be.data.label
+
+  @LuaFunction(mainThread = true)
+  fun setLabel(label: Optional<String>) {
+    be.data.label = label.map { it.take(MAX_LABEL_LENGTH) }.orElse(null)
+    be.dataUpdated()
+  }
+
+  @LuaFunction(mainThread = true)
+  fun getTooltip(): String? = be.data.tooltip
+
+  @LuaFunction(mainThread = true)
+  fun setTooltip(tooltip: Optional<String>) {
+    be.data.tooltip = tooltip.map { it.take(MAX_TOOLTIP_LENGTH) }.orElse(null)
+    be.dataUpdated()
+  }
+
+  @LuaFunction
+  fun setPaletteColor(index: Int, red: Int, green: Int, blue: Int) {
+    if (index !in 1..64) throw LuaException("Invalid palette index")
+    be.data.posterId = null
+    be.data.palette[index - 1] = (red shl 16) or (green shl 8) or blue
+    be.dataUpdated()
+  }
+
+  @LuaFunction
+  fun setPixel(x: Int, y: Int, color: Int) {
```

Comment:
These need to be on the main thread since they mutate the tile entity. This also means that each setPixel call will take at least 1 tick each, so see my comments on poster.lua for what could be done regarding bulk updates.

```suggestion
  @LuaFunction(mainThread = true)
  fun setPaletteColor(index: Int, red: Int, green: Int, blue: Int) {
    if (index !in 1..64) throw LuaException("Invalid palette index")
    be.data.posterId = null
    be.data.palette[index - 1] = (red shl 16) or (green shl 8) or blue
    be.dataUpdated()
  }

  @LuaFunction(mainThread = true)
  fun setPixel(x: Int, y: Int, color: Int) {
```